wave_containers/include/wave/containers/measurement_container.hpp

```diff
@@ -0,0 +1,313 @@
+#ifndef WAVE_CONTAINERS_MEASUREMENT_CONTAINER_HPP
+#define WAVE_CONTAINERS_MEASUREMENT_CONTAINER_HPP
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/composite_key.hpp>
+#include <boost/version.hpp>
+#include <iterator>
+#include <chrono>
+
+namespace wave {
+
+using TimeType = std::chrono::steady_clock::time_point;
+
+/** Internal implementation details - for developers only */
+namespace internal {
+
+using boost::multi_index::member;
+using boost::multi_index::indexed_by;
+using boost::multi_index::ordered_unique;
+using boost::multi_index::ordered_non_unique;
+using boost::multi_index::composite_key;
+using boost::multi_index::tag;
+
+/** To easily accomplish this container's goal - provide access to measurements
+ * sorted by time, by sensor, or both - we wrap a Boost.MultiIndex container.
+ *
+ * The multi_index_container template generates code allowing us to index the
+ * stored elements in different ways. Because these indices are specified at
+ * compile-time, the multi_index_container is configured using template
+ * parameters, which are defined here.
+ *
+ * See http://www.boost.org/doc/libs/1_63_0/libs/multi_index/doc/index.html for
+ * details on Boost.MultiIndex.
+ *
+ * The internal::measurement_container<T> holds all the type definitions
+ * required for a multi_index_container holding measurements of type T. Note
+ * this template is for convenience only, no objects are constructed.
+ */
+template<typename T>
+struct measurement_container {
+    // First, define which members of the Measurement object are used as keys
+    // Specify that the time key corresponds to the time_point member
+    struct time_key : member<T, TimeType, &T::time_point> {};
+    // Specify that the sensor key corresponds to the sensor_id member
+    struct sensor_key : member<T, decltype(T::sensor_id), &T::sensor_id> {};
+
+    // Define a composite key which combines the above two keys
+    // This lets us search elements sorted by sensor id first, then time
+    struct sensor_and_time_key : composite_key<T, sensor_key, time_key> {};
+
```

Comment:
Will it be easy to add additional keys for things like landmark ids, once we start using these to organize camera data?